FeedStoreChallenge/Core Data/CoreDataFeedStore.swift

```diff
@@ -0,0 +1,101 @@
+//
+//  CoreDataFeedStore.swift
+//  FeedStoreChallenge
+//
+//  Created by Wilmer Barrios on 23/01/21.
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import Foundation
+import CoreData
+
+public final class CoreDataFeedStore: FeedStore {
+	
+	typealias StoredFeedData = (feed: [CDFeedImageItem], timestamp: Date)
+	
+	private let context: NSManagedObjectContext
+	
+	public init(contextBuilder: CDContextBuilder) throws {
+		switch contextBuilder.build() {
+		case .success(let context):
+			self.context = context
+		case .failure(let e):
+			throw e
+		}
+	}
+	
+	// MARK: FeedStore extension
+	
+	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
+		let context = self.context
+		context.perform {
+			let fetchRequest: NSFetchRequest<NSFetchRequestResult> = CDFeedImage.fetchRequest()
+			let batchDeleteRequest = NSBatchDeleteRequest(fetchRequest: fetchRequest)
+			do {
+				try context.execute(batchDeleteRequest)
+				completion(nil)
+			} catch let e {
+				completion(e)
+			}
+		}
+	}
+	
+	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
+		let context = self.context
+		context.perform {
+			let feed_data: [CDFeedImageItem] = feed.toCD(context: context)
+			let storeFeed = CDFeedImage(context: context)
+			storeFeed.feed_data = NSOrderedSet(array: feed_data)
+			storeFeed.feed_timestamp = timestamp
+			do {
+				try context.save()
+				completion(nil)
+			} catch let e {
+				completion(e)
+			}
+		}
+	}
+	
+	public func retrieve(completion: @escaping RetrievalCompletion) {
+		let context = self.context
+		context.perform {
+			let fetchRequest: NSFetchRequest<CDFeedImage> = CDFeedImage.fetchRequest()
+			do {
+				let data = try context.fetch(fetchRequest)
+				guard let feed = data.last else { return completion(.empty) }
+				let (items, timestamp) = CoreDataFeedStore.map(feed)
+				completion(.found(feed: items, timestamp: timestamp))
```

Comment:
You can combine the guard checks into one, so you won't need the `map` function:

```suggestion
				let data = try context.fetch(fetchRequest)
				guard let feed = data.last, let items = feed.feed_data.array as? [CDFeedImageItem] else { return completion(.empty) }
				
				completion(.found(feed: items.toLocal(), timestamp: feed.feed_timestamp))
```